### PR TITLE
add CUB200 prototype datasets

### PR DESCRIPTION
**Pull request by pmeier**
This is the only popular vision dataset (according to [this list](https://paperswithcode.com/datasets?)) that we are currently missing.

cc @pmeier @bjuncek

### CODE REVIEW

**Comment by vadimkantorov** [comment]
will also fix this: https://github.com/pytorch/vision/issues/1654

**Comment by vadimkantorov** [comment]
Another frequent dataset for metric learning is Stanford Online Products...

File: torchvision/prototype/datasets/_builtin/cub.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        elif path.name == "train_test_split.txt":
+            return 1
+        elif path.name == "images.txt":
+            return 2
+        elif path.name == "bounding_boxes.txt":
+            return 3
+        else:
+            return None
+
+    _2011_SPLIT_MAP = {
```
Comment (NicolasHug) [inline torchvision/prototype/datasets/_builtin/cub.py:97]: Nit: maybe inline this below since we just use it once?

File: torchvision/prototype/datasets/_builtin/cub.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+            getitem(0),
+            buffer_size=INFINITE_BUFFER_SIZE,
+        )
+        return Mapper(dp, functools.partial(self._collate_and_decode_sample, year=config.year, decoder=decoder))
+
+    def _generate_categories(self, root: pathlib.Path) -> List[str]:
+        config = self.info.make_config(year="2011")
+        resources = self.resources(config)
+
+        dp = resources[0].load(pathlib.Path(root) / self.name)
```
Comment (NicolasHug) [inline torchvision/prototype/datasets/_builtin/cub.py:229]: Similarly to the changes in https://github.com/pytorch/vision/pull/5188, should this just be

```suggestion
        dp = resources[0].load(root)
```
?

File: torchvision/prototype/datasets/_builtin/cub.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        self, data: Tuple[str, Tuple[str, io.IOBase]], *, decoder: Optional[Callable[[io.IOBase], torch.Tensor]]
+    ) -> Dict[str, Any]:
+        _, (path, buffer) = data
+        content = read_mat(buffer)
+        return dict(
+            ann_path=path,
+            bounding_box=BoundingBox(
+                [int(content["bbox"][coord]) for coord in ("left", "bottom", "right", "top")], format="xyxy"
+            ),
+            segmentation=Feature(content["seg"]),
```
Comment (NicolasHug) [inline torchvision/prototype/datasets/_builtin/cub.py:141]: Could you help me understand why we use the `decoder` for the `Feature(...)` in `_2011_decode_ann`, but not in this function?

**Review by NicolasHug — APPROVED** [review]
Thanks @pmeier , some minor comments / questions but LGTM

File: torchvision/prototype/datasets/_builtin/cub.py (hunk 2 above)
Comment (pmeier) [reply torchvision/prototype/datasets/_builtin/cub.py:229]: Yes, good catch.

File: torchvision/prototype/datasets/_builtin/cub.py (hunk 3 above)
Comment (pmeier) [reply torchvision/prototype/datasets/_builtin/cub.py:141]: The `decoder` is used to turn an open file handle into a the pixel values as tensor. In this case the raw pixel values are already present as a `numpy.ndarray` so there is nothing left to decode.

**Comment by pmeier** [comment]
Test failure is unrelated.

**Comment by vadimkantorov** [comment]
I guess it would be nice to include some okay metric learning tutorial / reference impl using CUB/Stanford Online Products, this could be a useful example of new-style dataset usage
